### DISCUSSION TITLE
Quest 2: Retain VR session when switching between VR experience

### DISCUSSION THREAD

**Discussion by truonggiangdao** [pmndrs/xr#159] [Q&A]
Hi everyone, please advice: I'm trying to retain the VR session when user switch between 1 VR experience to another VR experience. 

Our use case is we have a existing exported virtual tour (old) that have VR support, now we try to see if it's possible to create another VR tour (new) using react-xr and let user navigate from old VR to new VR. The code below works on browser using WebXR Simulator Extension but when I try on Quest 2 everything becomes black. I have to press Menu/Oculus button on controller to escape the VR mode then I see the page is not navigating.

Here is my current testing code:

```
const CanvasContent = () => {
    const { gl } = useThree();
    
    useEffect(() => {
        if (window.myXRSession) {
            gl.xr.setSession(window.myXRSession);
        }
    }, [gl]);

    return (
        <Sphere image={myImage} />
    );
}

export const PageNew = () => {
    return (
        <VRCanvas>
           <ambientLight />
           <DefaultXRControllers />
            <CanvasContent />
        </VRCanvas>
    );
};
```

I'm using react-router-dom to navigate from old page to new Page

```
export const PageOld = () => {
    let navigate = useNavigate();
  
    function goToNewVR() {
      navigate("/page-new");
    }

    return (
        <OldTour
            vr="true"
            onVRSessionStarted={session => {window.myXRSession = session;}}
            onInteraction={goToNewVR}
        />
    );
};
```

Thanks in advanced!

**Edit:** I wondering if this is because the user was stuck in the old canvas even when it was destroyed. If that's the case, can the VRCanvas use the existing canvas instead of creating new canvas?

**Comment by CodyJasonBennett** [comment]
This isn't easily possible with three. There's more to the session init than requesting it, you have to bind it to your canvas via layers/session state and I don't believe that's readily exposed.

You could likely do this with a custom XR renderer (built on top of WebGLRenderer), as I outline here:
 - https://twitter.com/Cody_J_Bennett/status/1515269756647002114

**Comment by truonggiangdao** [comment]
@CodyJasonBennett, thanks for the advice. I'll try the custom XR renderer approach.
